### PR TITLE
Add `ddp_*_find_unused_parameters_false` to Plugins Registry.

### LINKED ISSUE #7744: Using DDPPlugin changes accelerator to ddp
## 🐛 Bug

If `DDPPlugin` is passed to `pl.Trainer` 

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1XsSQoQghHhIxA-9al9dnsjbBJOxzo1Ml?usp=sharing

### To Reproduce

As can be seen in output of "Run Test", `ddp` is initialized even though `dp` is chosen as accelerator.

### Expected behavior

Either ignore the plugin or raise an error about incompatability.

### Environment

* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu101
	- pytorch-lightning: 1.3.3
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Tue Apr 20 19:55:43 PDT 2021

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #7744 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)
- [ ] Did you list all the breaking changes introduced by this pull request?

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source/guides/speed.rst [hunk 1]
```diff
@@ -76,8 +76,8 @@ Prefer DDP over DP
 Whereas :class:`~pytorch_lightning.plugins.training_type.DDPPlugin` only performs 1 transfer to sync gradients, making DDP MUCH faster than DP.
 
 
-When using DDP set find_unused_parameters=False
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When using DDP or DDPSpawn, set find_unused_parameters=False
```
Comment (awaelchli) [inline docs/source/guides/speed.rst:79]: the term DDP already covered the family of ddp plugins. spawn is not the only one, there are others. 
so I think a simple note that this applies to all plugins supporting the `find_unused_parameters` would suffice

File: tests/plugins/test_plugins_registry.py [hunk 2]
```diff
@@ -107,3 +113,19 @@ def test_tpu_spawn_debug_plugins_registry(tmpdir):
     trainer = Trainer(plugins=plugin)
 
     assert isinstance(trainer.training_type_plugin, TPUSpawnPlugin)
+
+
+def test_ddp_spawn_training_type_plugins_registry(tmpdir):
```
Comment (awaelchli) [inline tests/plugins/test_plugins_registry.py:118]: we could simply parameterize this test with the name + corresponding expected plugin, e.g, these cases:

```python
("ddp_find_unused_parameters_false", DDPPlugin)
("ddp_spawn_find_unused_parameters_false", DDPSpawnPlugin)
("deepspeed_find_unused_parameters_false", DeepspeedPlugin)
```

and sharded etc. 
What do you think?

File: docs/source/guides/speed.rst [hunk 3]
```diff
@@ -76,8 +76,8 @@ Prefer DDP over DP
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (carmocca) [inline docs/source/guides/speed.rst:80]: Shouldn't this section also mention the availability of the `str`? Or do we mention it somewhere else?

File: tests/plugins/test_plugins_registry.py (hunk 2 above)
Comment (kaushikb11) [reply tests/plugins/test_plugins_registry.py:118]: Ofcourse.

File: docs/source/guides/speed.rst (hunk 1 above)
Comment (kaushikb11) [reply docs/source/guides/speed.rst]: Updated!

File: tests/plugins/test_plugins_registry.py (hunk 2 above)
Comment (kaushikb11) [reply tests/plugins/test_plugins_registry.py]: Updated!

File: docs/source/guides/speed.rst (hunk 3 above)
Comment (kaushikb11) [reply docs/source/guides/speed.rst]: `PluginsRegistry` will have a separate section with multiple DeepSpeed aliases and this.

File: docs/source/guides/speed.rst [hunk 4]
```diff
@@ -76,11 +76,14 @@ Prefer DDP over DP
...
 
 
-When using DDP set find_unused_parameters=False
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When using DDP plugins, set find_unused_parameters=False
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 By default we have set ``find_unused_parameters`` to True for compatibility issues that have arisen in the past (see the `discussion <https://github.com/PyTorchLightning/pytorch-lightning/discussions/6219>`_ for more information).
 This by default comes with a performance hit, and can be disabled in most cases.
 
+.. note::
```
Comment (carmocca) [inline docs/source/guides/speed.rst:84]: We always use notes :D

```suggestion
.. tip::
```

**Review by carmocca — APPROVED** [review]

File: docs/source/guides/speed.rst (hunk 1 above)
Comment (awaelchli) [reply docs/source/guides/speed.rst]: love it

**Review by awaelchli — APPROVED** [review]

File: docs/source/guides/speed.rst [hunk 5]
```diff
@@ -76,11 +76,14 @@ Prefer DDP over DP
 Whereas :class:`~pytorch_lightning.plugins.training_type.DDPPlugin` only performs 1 transfer to sync gradients, making DDP MUCH faster than DP.
 
 
-When using DDP set find_unused_parameters=False
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+When using DDP plugins, set find_unused_parameters=False
```
Comment (Borda) [inline docs/source/guides/speed.rst:79]:
```suggestion
When using DDP plugins, set ``find_unused_parameters=False``
```
maybe?

File: docs/source/guides/speed.rst [hunk 6]
```diff
@@ -76,11 +76,14 @@ Prefer DDP over DP
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 By default we have set ``find_unused_parameters`` to True for compatibility issues that have arisen in the past (see the `discussion <https://github.com/PyTorchLightning/pytorch-lightning/discussions/6219>`_ for more information).
```
Comment (Borda) [inline docs/source/guides/speed.rst:81]:
```suggestion
By default we have set ``find_unused_parameters`` to True for compatibility reasons that have been observed in the past (see the `discussion <https://github.com/PyTorchLightning/pytorch-lightning/discussions/6219>`_ for more details).
```

File: docs/source/guides/speed.rst [hunk 7]
```diff
@@ -76,11 +76,14 @@ Prefer DDP over DP
...
 This by default comes with a performance hit, and can be disabled in most cases.
```
Comment (Borda) [inline docs/source/guides/speed.rst:82]: any info how big is this hit, in %?

File: docs/source/guides/speed.rst [hunk 8]
```diff
@@ -90,6 +93,15 @@ This by default comes with a performance hit, and can be disabled in most cases.
         plugins=DDPPlugin(find_unused_parameters=False),
     )
 
+.. code-block:: python
```
Comment (Borda) [inline docs/source/guides/speed.rst:96]:
```suggestion
.. code-block::
```
the after `::` is a label not language type

File: docs/source/guides/speed.rst [hunk 9]
```diff
@@ -90,6 +93,15 @@ This by default comes with a performance hit, and can be disabled in most cases.
...
+
+    from pytorch_lightning.plugins import DDPSpawnPlugin
+
+    trainer = pl.Trainer(
+        gpus=2,
+        plugins=DDPSpawnPlugin(find_unused_parameters=False),
+    )
```
Comment (Borda) [inline docs/source/guides/speed.rst:103]: this can be `testcode` rather, right?

File: pytorch_lightning/plugins/training_type/sharded.py [hunk 10]
```diff
@@ -98,3 +98,12 @@ def pre_backward(self, closure_loss: torch.Tensor) -> None:
 
     def post_training_step(self):
         pass
+
+    @classmethod
+    def register_plugins(cls, plugin_registry: Dict):
```
Comment (Borda) [inline pytorch_lightning/plugins/training_type/sharded.py:103]: return type?

File: docs/source/guides/speed.rst (hunk 9 above)
Comment (kaushikb11) [reply docs/source/guides/speed.rst:103]: An example instead, we have internal tests.

File: docs/source/guides/speed.rst (hunk 7 above)
Comment (kaushikb11) [reply docs/source/guides/speed.rst:82]: I don't get what you mean?

File: pytorch_lightning/plugins/training_type/sharded.py (hunk 10 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/sharded.py]: Updated!

File: docs/source/guides/speed.rst [hunk 11]
```diff
@@ -76,11 +76,14 @@ Prefer DDP over DP
 Whereas :class:`~pytorch_lightning.plugins.training_type.DDPPlugin` only performs 1 transfer to sync gradients, making DDP MUCH faster than DP.
 
 
-When using DDP set find_unused_parameters=False
-^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-By default we have set ``find_unused_parameters`` to True for compatibility issues that have arisen in the past (see the `discussion <https://github.com/PyTorchLightning/pytorch-lightning/discussions/6219>`_ for more information).
+When using DDP plugins, set ``find_unused_parameters=False``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (awaelchli) [inline docs/source/guides/speed.rst:80]: Please undo the ticks or otherwise the title will not display correctly: 

image

**Review by awaelchli — CHANGES_REQUESTED** [review]

File: docs/source/guides/speed.rst (hunk 11 above)
Comment (awaelchli) [inline docs/source/guides/speed.rst:80]:
```suggestion
When using DDP plugins, set find_unused_parameters=False
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (awaelchli) [reply docs/source/guides/speed.rst:80]:
```suggestion
When using DDP plugins, set find_unused_parameters=False
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Review by awaelchli — APPROVED** [review]
😃 LGTM

File: docs/source/guides/speed.rst (hunk 9 above)
Comment (Borda) [reply docs/source/guides/speed.rst:103]: yes but rather keep all examples aligned with api

Comment (kaushikb11) [reply docs/source/guides/speed.rst:103]: `PluginsRegistry` will have a separate section with multiple DeepSpeed aliases and this. Will add it.

File: docs/source/guides/speed.rst [hunk 12]
```diff
@@ -90,6 +93,15 @@ This by default comes with a performance hit, and can be disabled in most cases.
         plugins=DDPPlugin(find_unused_parameters=False),
     )
 
+.. code-block::
```
Comment (Borda) [inline docs/source/guides/speed.rst:96]:
```suggestion
.. testcode::
```

File: docs/source/guides/speed.rst (hunk 7 above)
Comment (Borda) [reply docs/source/guides/speed.rst:82]: > comes with a performance hit

so how big is the hit?

**Review by Borda — APPROVED** [review]

File: docs/source/guides/speed.rst (hunk 7 above)
Comment (kaushikb11) [reply docs/source/guides/speed.rst:82]: >traversing the autograd graph introduces extra overheads, so applications should only set find_unused_parameters to True when necessary.
